src/main/java/org/logstash/beats/Message.java

```diff
@@ -1,27 +1,65 @@
 package org.logstash.beats;
 
-import java.util.HashMap;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.module.afterburner.AfterburnerModule;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufInputStream;
+
+import java.io.IOException;
+import java.io.InputStream;
 import java.util.Map;
 
 public class Message implements Comparable<Message> {
     private final int sequence;
     private String identityStream;
-    private final Map data;
+    private Map data;
     private Batch batch;
+    private ByteBuf buffer;
+
+    public final static ObjectMapper MAPPER = new ObjectMapper().registerModule(new AfterburnerModule());
 
+    /**
+     * Create a message using a map of key, value pairs
+     * @param sequence sequence number of the message
+     * @param map key/value pairs representing the message
+     */
     public Message(int sequence, Map map) {
         this.sequence = sequence;
         this.data = map;
+    }
 
-        identityStream = extractIdentityStream();
+    /**
+     * Create a message using a ByteBuf holding a Json object.
+     * Note that this ctr is *lazy* - it will not deserialize the Json object until it is needed.
+     * @param sequence sequence number of the message
+     * @param buffer {@link ByteBuf} buffer containing Json object
+     */
+    public Message(int sequence, ByteBuf buffer){
+        this.sequence = sequence;
+        this.buffer = buffer;
     }
 
+    /**
+     * Returns the sequence number of this messsage
+     * @return
+     */
     public int getSequence() {
         return sequence;
     }
 
-
-    public Map getData() {
+    /**
+     * Returns a list of key/value pairs representing the contents of the message.
+     * Note that this method is lazy if the Message was created using a {@ByteBuf}
+     * @return {@link Map} Map of key/value pairs
+     */
+    public Map getData(){
+        if (data == null && buffer != null){
```

Comment:
@robbavey don't we have to release this `ByteBuf` after we've deserialized its contents?

Reply:
This is a `slice`d from the Batch, which has its own separate `release` method to be called after it has been processed

Reply:
@robbavey ah right thanks :)

Reply:
@robbavey I think you could just `null` out `data` here though once it's not used anymore i.e. after deserialization. It's a minor win but it also makes the intention behind lazy deserialization more visible imo

Reply:
`getData` is called more than once - doesn't it make sense to keep it around for subsequent access?

Reply:
@robbavey sorry my bad. I meant null out `buffer` since we don't use that more than once :)

Reply:
@original-brownbear 👍 That makes a lot more sense ;)